EightMinuteEmpireLegends/Map/Map.cpp

```diff
@@ -0,0 +1,243 @@
+	return this->name;
+}
+
+int Map::numVertices() {
+	return this->v_vertices.size();
+}
+
+std::vector<Vertex*> Map::vertices() {
+	return this->v_vertices;
+}
+
+int Map::numEdges() {
+	return this->v_edges.size();
+}
+
+std::vector<Edge*> Map::edges() {
+	return this->v_edges;
+}
+
+int Map::numContinents() {
+	return this->v_continents.size();
+}
+
+std::vector<Continent*> Map::continents() {
+	return this->v_continents;
+}
+
+Edge* Map::getEdge(Vertex* v1, Vertex* v2) {
+	for (Edge* e : this->v_edges) {
+		if (e->getEndpoints()[0]->compare(v1) && e->getEndpoints()[1]->compare(v2)
+			|| e->getEndpoints()[0]->compare(v2) && e->getEndpoints()[1]->compare(v1)) {
+			return e;
+		}
+	}
+	return nullptr;
+}
+
+std::vector<Vertex*> Map::endVertices(Edge* e) {
+	return e != nullptr ? e->getEndpoints() : std::vector<Vertex*>();	
+}
+
+// Dangling pointer warning. Should fix in the future.
+Vertex* Map::opposite(Vertex* v, Edge* e) {
+	if (e != nullptr) {
+		if (e->getEndpoints()[0]->compare(v)) {
+			return e->getEndpoints()[1];
+		}
+		else if (e->getEndpoints()[1]->compare(v)) {
+			return e->getEndpoints()[0];
+		}
+	}	
+	return nullptr;
+}
+
+int Map::degree(Vertex* v) {
+	int degree = 0;
+	for (Edge* e : this->v_edges) {
+		if (e->getEndpoints()[0]->compare(v) || e->getEndpoints()[1]->compare(v)) {
+			degree++;
+		}
+	}
+	return degree;
+}
+
+std::vector<Edge*> Map::edges(Vertex* v) {
+	std::vector<Edge*> incidents{};
+	for (Edge* e : this->v_edges) {
+		if (e->getEndpoints()[0]->compare(v) || e->getEndpoints()[1]->compare(v)) {
+			incidents.push_back(e);
+		}
+	}
+	return incidents;
+}
+
+std::vector<Vertex*> Map::vertices(Continent* c) {
+	std::vector<Vertex*> regions{};
+
+	for (Vertex* v : this->v_vertices) {
+		if (v->getTerritory()->getContinent() == c->name) {
+			regions.push_back(v);
+		}
+	}
+	return regions;
+}
+
+bool Map::insertVertex(Vertex* v) {
+	if (exists(v->getTerritory()->getContinent())) {
+		for (Vertex* ver : this->v_vertices) {
+			if (ver->getId() == v->getId()) {
+				return false;
+			}
+		}
+		this->v_vertices.push_back(v);
+		return true;
+	}	
+	return false;
+}
+
+bool Map::insertEdge(Edge* e) {
+	if (getEdge(e->getEndpoints()[0], e->getEndpoints()[1]) == nullptr) {
+		this->v_edges.push_back(e);
+		return true;
+	}
+	return false;
+}
+
+bool Map::removeVertex(std::string id) {
+	std::vector<Vertex*>::iterator it = this->v_vertices.begin();
+	for (Vertex* v : this->v_vertices) {
+		if (v->getId() == id) {
+			this->v_vertices.erase(it);
+			removeEdgeCatalyst(v->getId());					
+			return true;
+		}
+		it++;
+	}
+	return false;
+}
+
+// Removes all incident edges of a deleted vertex.
+void Map::removeEdgeCatalyst(std::string id) {
+	std::vector<Edge*>::iterator begin = this->v_edges.begin();
+	std::vector<Edge*>::iterator end = this->v_edges.end();
+
+	this->v_edges.erase(std::remove_if(begin, end, [id](Edge* const e) { return (e->getEndpoints()[0]->getId() == id || e->getEndpoints()[1]->getId() == id); }), end);
```

Comment:
Please make this line more readable.

Reply:
Do you want me to add a comment? remove_if takes a lambda function as one of its arguments, which is what that terribly long line of code is. The lambda function returns true if the edge in the list is connected to the vertex being deleted.

Reply:
we can leave it like this. But personally I'm not a fan of fancy code like this. But up to you, it might be the preferred way.

Reply:
Maybe just break it up onto more than one line.
It's a little long.